File: samples/kmeans/cpp/src/kmeans.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

int main(int argc, char *argv[])
{
    cv::VideoCapture cap;
    cap.open(0);

    cv::Ptr<cv::FeatureDetector> orb = cv::ORB::create(500, 1.2, 8, 31, 0, 2, cv::ORB::HARRIS_SCORE, 31, 20);

    int kmeans_attempts = 10;
    int kmeans_K = 9;

    while (true)
    {
        cv::Mat frame, gray;
        cap >> frame;
        if (frame.empty())
        {
            break;
        }
        cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);

        std::vector<cv::KeyPoint> keypointVector;
        orb->detect(gray, keypointVector);
        if (keypointVector.size() <= kmeans_attempts)
        {
            continue;
        }

        cv::Mat kpMat(keypointVector.size(), 1, CV_32FC2);
        
        for (int i = 0; i < keypointVector.size(); i++)
        {
            cv::KeyPoint kp = keypointVector[i];
            kpMat.at<cv::Point2f>(i) = cv::Point2f(kp.pt.x, kp.pt.y);
        }

        // keypoints clustering with k-means
		cv::Mat bestLabel, center;
        cv::kmeans(kpMat, kmeans_K, bestLabel,
            cv::TermCriteria(cv::TermCriteria::EPS + cv::TermCriteria::COUNT, 10, 1.0),
            kmeans_attempts, cv::KMEANS_PP_CENTERS, center);

        // draw keypoints
        cv::drawKeypoints(frame, keypointVector, frame);

        // draw kmeans centroid points
        cv::Scalar color(255, 0, 255, 0);
        for (int i = 0; i < center.rows; i++)
        {
            cv::Point2f cp = center.at<cv::Point2f>(i);
            cv::circle(frame, cp, 6, color, cv::FILLED);
            cv::circle(frame, cp, 3, cv::Scalar::all(255), cv::FILLED);
            cv::circle(frame, cp, 1, color, cv::FILLED);
        }

        cv::imshow("KMeans", frame);
        int key = cv::waitKey(10);
        if (key == 27)
        { // ESC to exit
            break;
        }
    }
    cap.release();
    cv::destroyAllWindows();
    return 0;
}

File: samples/kmeans/java/KMeans.java
import org.bytedeco.javacpp.indexer.FloatIndexer;
import org.bytedeco.opencv.global.opencv_core;
import org.bytedeco.opencv.global.opencv_features2d;
import org.bytedeco.opencv.global.opencv_imgproc;
import org.bytedeco.opencv.opencv_core.*;
import org.bytedeco.opencv.opencv_features2d.ORB;
import org.bytedeco.opencv.opencv_videoio.VideoCapture;

public class KMeans {
    public static void main(String[] args) {
        VideoCapture cap = new VideoCapture(0);

        ORB orb = ORB.create(500, 1.2f, 8, 31, 0, 2, ORB.HARRIS_SCORE, 31, 20);
        KeyPointVector keyPointVector = new KeyPointVector();
        Mat grayMat = new Mat();
        Mat bestLabel = new Mat();
        Mat center = new Mat();

        int kmeans_attempts = 10;
        int kmeans_K = 9;

        while (true) {
            Mat frame = new Mat();
            boolean ret = cap.read(frame);
            if (!ret) break;

            opencv_imgproc.cvtColor(frame, grayMat, opencv_imgproc.COLOR_BGR2GRAY);
            orb.detect(grayMat, keyPointVector);
            if (keyPointVector.size() <= kmeans_K) {
                continue;
            }

            Mat keypointMat = new Mat((int) keyPointVector.size(), 1, opencv_core.CV_32FC2);
            FloatIndexer kpMatIdx = keypointMat.createIndexer();
            for (int i = 0; i < keypointMat.rows(); i++) {
                KeyPoint keyPoint = keyPointVector.get(i);
                kpMatIdx.put(i, keyPoint.pt().x(), keyPoint.pt().y());
            }
            // keypoints clustering with k-means
            opencv_core.kmeans(keypointMat, kmeans_K, bestLabel,
                    new TermCriteria(TermCriteria.EPS + TermCriteria.COUNT, 10, 1.0),
                    kmeans_attempts, opencv_core.KMEANS_PP_CENTERS, center);

            // draw keypoints
            opencv_features2d.drawKeypoints(frame, keyPointVector, frame);

            // draw kmeans centroid points
            FloatIndexer centerIdx = center.createIndexer();
            Scalar color = new Scalar(255, 0, 255, 0);
            for (int i = 0; i < center.rows(); i++) {
                int cx = (int) centerIdx.get(i, 0);
                int cy = (int) centerIdx.get(i, 1);
                opencv_imgproc.circle(frame, new Point(cx, cy), 6, color, opencv_imgproc.FILLED, opencv_imgproc.CV_AA, 0);
                opencv_imgproc.circle(frame, new Point(cx, cy), 3, Scalar.all(255), opencv_imgproc.FILLED, opencv_imgproc.CV_AA, 0);
                opencv_imgproc.circle(frame, new Point(cx, cy), 1, color, opencv_imgproc.FILLED, opencv_imgproc.CV_AA, 0);
            }

            OpenCVFX.imshow("KMeans", frame);
            if (OpenCVFX.waitKey() == 27) {
                break;
            }

            releaseMat(frame, keypointMat);
        }
        OpenCVFX.destroyAllWindows();
        cap.release();
    }

    private static void releaseMat(Mat... mats) {
        for (Mat mat : mats) {
            if (mat != null) {
                mat.release();
                mat.close();
            }
        }
    }
}
